### PR TITLE
Fix #6403: Do not intercept crashes

### LINKED ISSUE #6403: Fuzzing test succeeds even though the compiler crashes
```shell
cp tests/pending/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala
```
```scala
> testCompilation tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala

[warn] Multiple main classes detected.  Run show discoveredMainClasses to see the list
[info] Test run started
[info] Test dotty.tools.dotc.CompilationTests.negAll started
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
[info] Test dotty.tools.dotc.CompilationTests.runAll started
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
[info] Test dotty.tools.dotc.CompilationTests.pickling started
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
[info] Test dotty.tools.dotc.CompilationTests.pos started
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
[info] Test dotty.tools.dotc.CompilationTests.tastyBootstrap started
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
[info] Test dotty.tools.dotc.CompilationTests.posTwice started
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test
[info] Test dotty.tools.dotc.CompilationTests.fuzzyAll started
java.lang.StackOverflowError            ] completed (0/1, 0 failed, 3s)
        at java.util.IdentityHashMap.remove(IdentityHashMap.java:520)
        at dotty.tools.dotc.core.SymDenotations$ClassDenotation.recur$1(SymDenotations.scala:1860)
        at dotty.tools.dotc.core.SymDenotations$ClassDenotation.baseTypeOf(SymDenotations.scala:1867)
        at dotty.tools.dotc.core.Types$Type.baseType(Types.scala:925)
        at dotty.tools.dotc.core.Types$NamedType.argForParam(Types.scala:2060)
        at dotty.tools.dotc.core.Types$NamedType.derivedSelect(Types.scala:2112)
        at dotty.tools.dotc.core.Types$TypeMap.derivedSelect(Types.scala:4345)
        at dotty.tools.dotc.core.Types$ApproximatingTypeMap.derivedSelect(Types.scala:4637)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:63)
        at dotty.tools.dotc.core.Types$TypeMap.mapArgs$1(Types.scala:4407)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:4415)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:69)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:4437)
        at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:69)
        at dotty.tools.dotc.core.TypeOps.asSeenFrom(TypeOps.scala:30)
        at dotty.tools.dotc.core.Contexts$Context.asSeenFrom(Contexts.scala:71)
        at dotty.tools.dotc.core.Types$Type.asSeenFrom(Types.scala:832)
        at dotty.tools.dotc.core.Types$Type.memberInfo(Types.scala:825)
        at dotty.tools.dotc.transform.GenericSignatures$.jsig$1(GenericSignatures.scala:219)
        at dotty.tools.dotc.transform.GenericSignatures$.boxedSig$1(GenericSignatures.scala:64)
        at dotty.tools.dotc.transform.GenericSignatures$.argSig$1(GenericSignatures.scala:145)
        at dotty.tools.dotc.transform.GenericSignatures$.classSig$1$$anonfun$1(GenericSignatures.scala:174)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at dotty.tools.dotc.transform.GenericSignatures$.classSig$2(GenericSignatures.scala:174)
        at dotty.tools.dotc.transform.GenericSignatures$.jsig$1(GenericSignatures.scala:228)
        at dotty.tools.dotc.transform.GenericSignatures$.boxedSig$1(GenericSignatures.scala:64)
        at dotty.tools.dotc.transform.GenericSignatures$.argSig$1(GenericSignatures.scala:145)
        at dotty.tools.dotc.transform.GenericSignatures$.classSig$1$$anonfun$1(GenericSignatures.scala:174)
        at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    
[...]
[=======================================] completed (1/1, 0 failed, 4s)
[info] Test dotty.tools.dotc.CompilationTests.genericJavaSignatures started
No files matched "tests/fuzzy/SOE-9cb9025fd025d2e548933c770ce32877f4e21b50.scala" in test

================================================================================
Test Report
================================================================================

1 suites passed, 0 failed, 1 total

[info] Test run finished: 0 failed, 0 ignored, 8 total, 5.452s
[info] Test run started
[info] Test run finished: 0 failed, 0 ignored, 0 total, 0.0s
[info] Passed: Total 8, Failed 0, Errors 0, Passed 8
[success] Total time: 8 s, completed 30 avr. 2019 15:54:00
```

### PR DESCRIPTION
**Pull request by nicolasstucki**
Emitting an error when the backend crashes hides the crash from the reporter and fails which cannot be identified by the testing framework.

### CODE REVIEW

File: compiler/src/dotty/tools/backend/jvm/GenBCode.scala [hunk 1]
```diff
@@ -180,8 +180,7 @@ class GenBCodePipeline(val entryPoints: List[Symbol], val int: DottyBackendInter
             try   { /*withCurrentUnit(item.cunit)*/(visit(item)) }
             catch {
               case ex: Throwable =>
-                ex.printStackTrace()
-                ctx.error(s"Error while emitting ${int.sourceFileFor(item.cunit)}\n${ex.getMessage}")
+                throw new Exception(s"Error while emitting ${int.sourceFileFor(item.cunit)}", ex)
```
Comment (smarter) [inline compiler/src/dotty/tools/backend/jvm/GenBCode.scala]: There's another similar case in https://github.com/lampepfl/dotty/blob/ab42b4d76f50f8efe90d1388b98ae2644b31903b/compiler/src/dotty/tools/backend/jvm/GenBCode.scala#L415

File: compiler/test/dotty/tools/vulpix/ParallelTesting.scala [hunk 2]
```diff
@@ -852,21 +852,21 @@ trait ParallelTesting extends RunnerOrchestration { self =>
...
-                case ex: Throwable => fail(s"Fatal compiler crash when compiling: ${testSource.title}")
-              }
-            if (reporter.compilerCrashed)
-              fail(s"Compiler crashed when compiling: ${testSource.title}")
+            try {
+              val reporter = compile(sourceFiles, flags, true, outDir)
+              if (reporter.compilerCrashed)
+                fail(s"Compiler crashed when compiling: ${testSource.title}")
+            }
+            catch {
```
Comment (allanrenucci) [inline compiler/test/dotty/tools/vulpix/ParallelTesting.scala]: It should not happen, right? The exception is already caught in `def compile`: https://github.com/lampepfl/dotty/blob/c25fbea02bb3620fbec6f142fc6cbbbfdbeda5ed/compiler/test/dotty/tools/vulpix/ParallelTesting.scala#L374-L390

Comment (nicolasstucki) [reply compiler/test/dotty/tools/vulpix/ParallelTesting.scala]: In theory it should not happen. But if it does, we would report it and not crash.

Comment (allanrenucci) [reply compiler/test/dotty/tools/vulpix/ParallelTesting.scala]: But if a crash happens, it is a bug in the test framework and a crash is arguably better than reporting a test failure.

Also, `compile` is called in multiple places in this file and that's the only place where it is wrapped in a try catch:
- https://github.com/lampepfl/dotty/blob/29aeb02be0c08a20373027a0d80deaa87cb699ff/compiler/test/dotty/tools/vulpix/ParallelTesting.scala#L589
- https://github.com/lampepfl/dotty/blob/29aeb02be0c08a20373027a0d80deaa87cb699ff/compiler/test/dotty/tools/vulpix/ParallelTesting.scala#L598
- https://github.com/lampepfl/dotty/blob/29aeb02be0c08a20373027a0d80deaa87cb699ff/compiler/test/dotty/tools/vulpix/ParallelTesting.scala#L685

Comment (smarter) [reply compiler/test/dotty/tools/vulpix/ParallelTesting.scala]: I agree with Allan

Comment (nicolasstucki) [reply compiler/test/dotty/tools/vulpix/ParallelTesting.scala]: Ok

File: compiler/src/dotty/tools/backend/jvm/GenBCode.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/backend/jvm/GenBCode.scala]: Modified

**Comment by nicolasstucki** [comment]
The printed message might be swallowed by the test framework. It feels more resilient to wrap them.

**Comment by nicolasstucki** [comment]
Maybe we should have both. I will create some `object CompilerCrash` with the logic that we should use for all compiler crashes. This way we can have a homogeneous behavior that we could modify easly.

**Comment by nicolasstucki** [comment]
Now it prints the extra message and throws the original exception

File: compiler/src/dotty/tools/backend/jvm/DottyBackendInterface.scala [hunk 3]
```diff
@@ -482,11 +482,19 @@ class DottyBackendInterface(outputDirectory: AbstractFile, val superCallsMap: Ma
...
   private def verifySignature(sym: Symbol, sig: String)(implicit ctx: Context): Unit = {
     import scala.tools.asm.util.CheckClassAdapter
-    def wrap(body: => Unit): Boolean =
-      try { body; true }
-      catch { case ex: Throwable => println(ex.getMessage); false }
+    def wrap(body: => Unit): Unit = {
+      try body
+      catch {
+        case ex: Throwable =>
+          println(i"""|compiler bug: created invalid generic signature for $sym in ${sym.denot.owner.showFullName}
```
Comment (smarter) [inline compiler/src/dotty/tools/backend/jvm/DottyBackendInterface.scala]: Using ctx.error would be better I think

**Pull request merged**